src/constructors/lattices.c

```diff
@@ -0,0 +1,308 @@
+ *        the vertex with the second coordinate equal to the index.
+ *
+ * \return Error code:
+ *         \c IGRAPH_EINVAL: invalid (negative) length of row_lengths_vector does not match the length of the
+            row_start_vector.
+ *
+ * Time complexity:  O(|V|), where |V| is the number of vertices in the generated graph.
+ */
+static igraph_error_t triangle_lattice(igraph_t *graph, igraph_bool_t directed, igraph_bool_t mutual,
+                                       const igraph_vector_int_t *row_lengths_vector, const igraph_vector_int_t *row_start_vector)
+{
+    igraph_vector_int_t edges = IGRAPH_VECTOR_NULL;
+    igraph_integer_t row_count = igraph_vector_int_size(row_lengths_vector);
+    igraph_integer_t no_of_nodes;
+    igraph_vector_int_t row_lengths_prefix_sum_vector;
+    igraph_integer_t i, j;
+
+    if (igraph_vector_int_size(row_lengths_vector) != igraph_vector_int_size(row_start_vector))
+    {
+        IGRAPH_ERRORF(
+            "Length of row_lengths_vector vector (%" IGRAPH_PRId ") must match the length of the "
+            "row_start_vector (%" IGRAPH_PRId ").",
+            IGRAPH_EINVAL,  igraph_vector_int_size(row_lengths_vector), igraph_vector_int_size(row_start_vector));
+    }
+
+    for (i = 0; i < row_count; i++)
+    {
+        if (VECTOR(*row_lengths_vector)[i] < 0)
+        {
+            IGRAPH_ERRORF(
+                "row_lengths_vector vector must have non-negative coordinates, "
+                "was (%" IGRAPH_PRId ") for the (%" IGRAPH_PRId ")-th row.",
+                IGRAPH_EINVAL, VECTOR(*row_lengths_vector)[i], i);
+        }
+    }
+
+    IGRAPH_VECTOR_INT_INIT_FINALLY(&edges, 0);
+
+    IGRAPH_CHECK(igraph_vector_int_init(&row_lengths_prefix_sum_vector, row_count + 1));
+    IGRAPH_FINALLY(igraph_vector_int_destroy, &row_lengths_prefix_sum_vector);
+
+    VECTOR(row_lengths_prefix_sum_vector)[0] = 0;
+    for (i = 1; i < row_count + 1; i++)
+    {
+        IGRAPH_SAFE_ADD(VECTOR(row_lengths_prefix_sum_vector)[i - 1], VECTOR(*row_lengths_vector)[i - 1], &(VECTOR(row_lengths_prefix_sum_vector)[i]));
+    }
+    no_of_nodes = VECTOR(row_lengths_prefix_sum_vector)[row_count];
+
+#define VERTEX_INDEX(i, j) (VECTOR(row_lengths_prefix_sum_vector)[j] + i - VECTOR(*row_start_vector)[j])
+#define ROW_END(j) (VECTOR(*row_start_vector)[j] + VECTOR(*row_lengths_vector)[j] - 1)
+#define ADD_EDGE_IJ_KL_IF_EXISTS(i, j, k, l)                                                                                                      \
+    if (VECTOR(*row_start_vector)[l] <= k && k <= ROW_END(l) && 0 <= l && l <= row_count - 1)                                                     \
+    {                                                                                                                                             \
+        igraph_vector_int_push_back(&edges, VERTEX_INDEX((i), (j)));                                                                              \
```

Comment:
Please add a comment either here or above the `ADD_EDGE_IJ_KL_IF_EXISTS` macro to explain that `&edges` is guaranteed to have enough space for the new items so that's why we don't need an `IGRAPH_CHECK(...)` call. We typically do this by adding `/* reserved */` or `/* already reserved */` to the end of the line calling `igraph_vector_int_push_back(...)`